paddle/fluid/framework/details/all_reduce_deps_pass.cc

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "paddle/fluid/framework/details/all_reduce_deps_pass.h"
+#include "paddle/fluid/framework/details/all_reduce_op_handle.h"
+#include "paddle/fluid/framework/details/multi_devices_helper.h"
+#include "paddle/fluid/framework/details/op_graph_view.h"
+#include "paddle/fluid/framework/details/var_handle.h"
+#include "paddle/fluid/framework/op_proto_maker.h"
+
+namespace paddle {
+namespace framework {
+namespace details {
+
+VarHandle* GetValidInput(const AllReduceOpHandle* a) {
+  for (auto p : a->Inputs()) {
+    VarHandle* b = dynamic_cast<VarHandle*>(p);
+    if (b) {
+      return b;
+    }
+  }
+
+  return nullptr;
+}
+
+struct SortNode {
+  SortNode(AllReduceOpHandle* op, std::unordered_map<std::string, int>* vars) {
+    op_ = op;
+    vars_ = vars;
+  }
+
+  AllReduceOpHandle* op_;
+  std::unordered_map<std::string, int>* vars_;
+
+  bool operator<(const SortNode& r) const {
+    VarHandle* i0 = dynamic_cast<VarHandle*>(GetValidInput(op_));
+    VarHandle* i1 = dynamic_cast<VarHandle*>(GetValidInput(r.op_));
+
+    PADDLE_ENFORCE(i0 != nullptr && i1 != nullptr, "%s convert to %s error",
+                   op_->DebugString(), r.op_->DebugString());
+
+    auto l_it = vars_->find(i0->name_);
+    auto r_it = vars_->find(i1->name_);
+
+    if (l_it->second < r_it->second) return true;
+
+    return false;
+  }
+};
+
+std::unique_ptr<ir::Graph> AllReduceDepsPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  constexpr char kAllOpDescs[] = "all_op_descs";
+  auto& graph_ops = ir::FilterByNodeWrapper<OpHandleBase>(*graph)
+
+      // get vars level
+      int level = 0;
+  std::unordered_map<std::string, int> vars;
+  auto& ops = Get<const std::vector<OpDesc*>>(kAllOpDescs);
+  for (auto* op_desc : ops) {
+    auto outputs = op_desc->Outputs();
+    for (auto& o_it : outputs) {
+      for (auto& v : o_it.second) {  // values
+        vars[v] = level;
+        level++;
+      }
+    }
+  }
+
+  std::unordered_map<AllReduceOpHandle*, int> allreduce_ops;
```

Comment:
Why `all reduce_ops` is the map? It seems that the mapped values are always zero. Maybe you need a set here?